autopush/router/apns2.py

```diff
@@ -0,0 +1,152 @@
+        :type metrics: autopush.metrics.IMetric
+        """
+        self.server = SERVER if not use_sandbox else SANDBOX
+        self.port = 443 if not alt else 2197
+        self.ssl_context = hyper.tls.init_context(cert=(cert_file, key_file))
+        self.log = logger
+        self.metrics = metrics
+        self.topic = topic
+        self._max_connections = max_connections
+        self._min_connections = min_connections
+        self.connections = deque(maxlen=max_connections)
+        if self.log:
+            self.log.info("Starting APNS connection")
+        # HTTP20Connection is not thread safe, so it's best to keep processing
+        # of the result in the same thread.
+        for i in xrange(0, min_connections):
+            self._get_connection()
+
+    def send(self, router_token, payload, apns_id=None,
+             priority=True, topic=None, exp=None):
+        """Send the dict of values to the remote bridge
+
+        This sends the raw data to the remote bridge application using the
+        APNS2 HTTP2 API.
+
+        :param router_token: APNs provided hex token identifying recipient
+        :type router_token: str
+        :param payload: Data to send to recipient
+        :type payload: dict
+        :param priority: True is high priority, false is low priority
+        :type priority: bool
+        :param topic: BundleID for the recipient application (overides default)
+        :type topic: str
+        :param exp: Message expiration timestamp
+        :type exp: timestamp
+
+        """
+        if not apns_id:
+            apns_id = str(uuid.uuid4()).lower()
+        body = json.dumps(payload)
+        if priority:
+            priority = APNS_PRIORITY_IMMEDIATE
+        else:
+            priority = APNS_PRIORITY_LOW
+        headers = {
+            'apns-id': apns_id,
+            'apns-priority': priority,
+            'apns-topic': topic or self.topic,
+        }
+        if exp:
+            headers['apns-expiration'] = exp
+        url = '/3/device/' + router_token
+        connection = self._get_connection()
+        d = deferToThread(
+            self._process_send,
+            connection=connection,
+            url=url,
+            body=body,
+            headers=headers,
+        )
+        # failed connections self-close
+        d.addCallback(self._return_connection)
+        return d
+
+    def _get_connection(self):
+        try:
+            return self.connections.pop()
+        except IndexError:
+            return HTTP20Connection(
+                self.server, self.port,
+                ssl_context=self.ssl_context,
+                force_proto='h2')
+
+    def _return_connection(self, connection):
+        if len(self.connections) < self._min_connections:
+            self.connections.appendleft(connection)
+        else:
+            connection.close()
+
+    def _process_send(self, connection, url, body, headers):
+        # HTTP20Connection is not thread safe. It's best to keep this and
```

Comment:
It can't be used concurrently. The docs note that if you're careful, you can use the stream_id identifier to allow it to be used, but still note caution should be applied. Using the same connection for the duration of a thread's requirement seems the safest method.